module/zstd/zstd.c

```diff
@@ -0,0 +1,622 @@
+static enum zio_zstd_levels
+zstd_cookie_to_enum(int32_t level)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fastlevels); i++) {
+		if (fastlevels[i].cookie == level) {
+			return (fastlevels[i].level);
+		}
+	}
+
+	/* This shouldn't happen. Fall back to the default level. */
+	printk(KERN_ERR "%s:Invalid ZSTD level encountered: %d",
+	    __func__, level);
+	return (ZIO_ZSTD_LEVEL_DEFAULT);
+}
+
+/* Convert zstd_level to internal stored zfs level */
+static int32_t
+zstd_enum_to_cookie(enum zio_zstd_levels elevel)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fastlevels); i++) {
+		if (fastlevels[i].level == elevel)
+			return (fastlevels[i].cookie);
+	}
+
+	/* This shouldn't happen. Fall back to the default level. */
+	printk(KERN_ERR "%s:Invalid ZSTD enum level encountered: %d",
+	    __func__, elevel);
+	return (ZIO_ZSTD_LEVEL_DEFAULT);
+}
+
+/* Compress block using zstd */
+size_t
+zstd_compress(void *s_start, void *d_start, size_t s_len, size_t d_len,
+    int level)
+{
+	size_t c_len;
+	uint32_t bufsize;
+	int32_t levelcookie;
+	char *dest = d_start;
+	ZSTD_CCtx *cctx;
+
+	levelcookie = zstd_enum_to_cookie(level);
+	ASSERT3U(d_len, >=, sizeof (bufsize));
+	ASSERT3U(d_len, <=, s_len);
+	ASSERT3U(levelcookie, !=, 0);
+
+	cctx = ZSTD_createCCtx_advanced(zstd_malloc);
+
+	/*
+	 * Out of kernel memory, gently fall through - this will disable
+	 * compression in zio_compress_data
+	 */
+	if (!cctx) {
+		return (s_len);
+	}
+
+	/* Set the compression level */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_compressionLevel, levelcookie);
+
+	/* Use the "magicless" zstd header which saves us 4 header bytes */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_format, ZSTD_f_zstd1_magicless);
+
+	/*
+	 * Disable redundant checksum calculation and content size storage since
+	 * this is already done by ZFS itself.
+	 */
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 0);
+	ZSTD_CCtx_setParameter(cctx, ZSTD_c_contentSizeFlag, 0);
+
+	c_len = ZSTD_compress2(cctx,
+	    &dest[sizeof (bufsize) + sizeof (levelcookie)],
+	    d_len - sizeof (bufsize) - sizeof (levelcookie),
+	    s_start, s_len);
+
+	ZSTD_freeCCtx(cctx);
+
+	/* Error in the compression routine, disable compression. */
+	if (ZSTD_isError(c_len)) {
+		return (s_len);
+	}
+
+	/*
+	 * Encode the compressed buffer size at the start. We'll need this in
+	 * decompression to counter the effects of padding which might be added
+	 * to the compressed buffer and which, if unhandled, would confuse the
+	 * hell out of our decompression function.
+	 */
+	bufsize = c_len;
+	*(uint32_t *)dest = BE_32(bufsize);
+
+	/*
+	 * Encode the compression level as well. We may need to know the
+	 * original compression level if compressed_arc is disabled, to match
+	 * the compression settings to write this block to the L2ARC. Encode
+	 * the actual level, so if the enum changes in the future, we will be
+	 * compatible.
+	 */
+	*(uint32_t *)(&dest[sizeof (bufsize)]) = BE_32(levelcookie);
```

Comment:
this is a redundant discussion. zstd has a own header magic. my co developers here removed it. we dont need to introduce a own magic, we simply need to turn on the zstd own magic again. i had the same oppinion as you, but people voted against me.

Reply:
@behlendorf have a look at the discussion, please https://github.com/openzfs/zfs/pull/10278#issuecomment-630089022

TL;DR: 

[zstd is considered to be stable](https://github.com/openzfs/zfs/pull/10278#issuecomment-630300535) and a change of the magic is unlikely as it would imply a change in format (i.e. it's not `zstd` anymore).

Personally I think this is very unlikely to happen, but even when there was a change in the future, we still could provide an upgrade path by checking for non-existence of the magic (old zstd) or existence (new "zstd") to give the decompressor a hint.

@Cyan4973 had another point: interoperability. There were cases where people got confused why decompression of headerless data wasn't possible with standard tools, which rely on that magic. There is no valid use case for using zstd tools on zfs blocks. Even when doing forensics, it's unlikely to use zstd standard tools, but one has to look deep into the stuff already and use customized tools/scripts/...

IMHO we don't really need it, but when you like to readd it, we can do

Reply:
In this case, it is less about the stability of zstd, as ZFS's requirement to be able to compress the same block and get the same checksum as a result. This cannot happen if ZSTD compresses the block differently (or better, as each new version often does). So we want to store something like the zstd ZSTD_VERSION_NUMBER define, so we can possibly embed more than 1 version of zstd if needed.

We can likely just use the top 16 bits of the 'compression level' cookie to store the version, and the lower 16 bits for the level.

Reply:
ah, got it! thx @allanjude 

Well, in this case the magic wouldn't help anything. But we definitely need some test so we know when we have to do any adjustments

Reply:
Shoot, thats a very good point @allanjude 
It make it even more redundant to also include the ZSTD magic and won't take more space. So it's a win-win scenario that way :)

Reply:
@allanjude did a much better job articulating this than I.  But yes, Allan made exactly the point I was trying to above.  Splitting the the cookie sounds like an entirely reasonable way to handle this.

Reply:
i agree. so lets reserve the upper 16 bits with it and lets start with version zero. so we dont need to change any piece of code. :-) just joking. i will at least specify the field and initialize it with 0

Reply:
use `ZSTD_VERSION_NUMBER` from `zstdlib.h`, please, so noone can forget to update it ;)

Reply:
@c0d3z3r0 i'm already waiting for the day zstd version 6 is released. zfs is a very long term project. never forget the year 2000 problem and what it caused and also remember the future 2035 problem. its not a good idea to start with a value of 10405. if zstd version 6 is reached you will remember my words

Reply:
what about using `ZSTD_VERSION_NUMBER - 10405` (which is 0) as a start point?

Reply:
wait ... forget it lol. I should go sleep now haha

Reply:
@c0d3z3r0 :-) damn i wanted already to comment something

Reply:
We could switch to a separate 32 bits for the version, but that might start to get out of hand

Reply:
true

Reply:
i found another small thing which we now need to deal. the cookie is signed. this makes now dealing with mask more interesting

Reply:
I'm not really happy with having to update the version on every new zstd release, but it should be enough to do it when an imcompatibility occurrs, right?

Reply:
"these 2 functions" - the mapping functions? I don't see how that is possible, due to the fast algos being negative

Reply:
ah, you want to store the enum instead of the zstd level, but keep them for compression/decompression?

Reply:
@allanjude have a look at the latest commit, please

Reply:
done dcd060, f23e8, e90dd4

Reply:
extended to 24 bits now -> max. version 1677.72.15, I'd say that should be more than enough